synapse/federation/federation_server.py

```diff
@@ -843,8 +843,22 @@ async def _on_send_membership_event(
                 Codes.BAD_JSON,
             )
 
+        # Note that get_room_version throws if the room does not exist here.
         room_version = await self.store.get_room_version(room_id)
 
+        if await self.store.is_partial_state_room(room_id):
+            # If our server is still only partially joined, we can't give a complete
+            # response to send_join, so return a 404 as we would if we weren't in the
+            # room at all.
+            logger.info(
+                "Rejecting send_join to %s because it's a partial state room", room_id
+            )
+            raise SynapseError(
+                404,
+                "Unable to handle send_join right now; this server is not fully joined.",
+                errcode=Codes.NOT_FOUND,
+            )
```

Comment:
We also handle `/send_knock` and `/send_leave` requests here, so the log and error messages will be misleading for those cases.